### DISCUSSION TITLE
Quality calculated,from luminance and chrominance

### DISCUSSION THREAD

**Discussion by FredWahl** [ImageMagick/ImageMagick#7384] [General]
im writing a small utility that automates cjpeg such that some pictures can be compressed. I don’t want to run it on files where this quality value is 75 or smaller and also files that already are small. This value must be calculated. With some difficulty, I managed to get some quality value from imagemagick which turned out to be compression quality which I don’t even know what it is. If thare’s a way to get this value, i would like to know how. If not, I would like to make it a request which I think would be of general interest. It seems as if the exiftool can give me this value using exiftool -JPEGQualityEstimate yourimage.jpg.

**Comment by snibgo** [comment]
I make x.jpg with `-quality 41`. Then:

```
D:\temp>exiftool -JPEGQualityEstimate x.jpg
JPEG Quality Estimate           : 41

D:\temp>%IMG7%magick x.jpg -verbose info: |grep -i quality
  Quality: 41

D:\temp>%IMG7%magick x.jpg -format %Q\n info:
41
```
Does that answer your question?

**Reply by fmw42** [reply]
Also

`magick x.jpg -format "%[quality]\n" info`:

See %[quality] at https://imagemagick.org/script/escape.php

**Reply by FredWahl** [reply]
Thank You! I will try that tomorrow because it’s late. I feel like I have climbed a high mountain only to find I climbed the wrong mountain. That’s what I felt when I found out my code gave thae same value after compression.

**Reply by FredWahl** [reply]
I read the link and found this
%Q	image compression quality ( 0 = default )
%[quality]	Image quality value (as of IM 7.0.3.8)
for some reason, I still get 93. I expected 75 since the image I used had been compressed with cjpeg using 75 as the quality parameter. I need to do some more testing

**Reply by FredWahl** [reply]
Thanks snibgo. I think this command will work because exiftool only extracts this kind of data, i think. I use windows so the grep command doesn’t work. I tried verbose and it was really slow.

**Comment by snibgo** [comment]
Programs that compress images with the JPEG method often use a "quality" number between 1 and 100, where 1 is the worst quality but greatest compression.

The meaning of the quality number is not defined in standards, so one program's quality "75" might result in a visually lower quality than another program's. In addition, the quality number is not written to the file as metadata. So any program that reads a JPEG image and gives a quality number is really *estimating* the number. If the same library was used to write and read the JPEG, the estimated number is likely to be consistent, but another library might give a different estimate.

`grep` works on Windows, if you have that program installed. Windows comes with `findstr`, which is similar.

With IM, you can use `-ping` to avoid processing the pixels. The estimation doesn't need pixel data. For example:
```
magick -ping in.jpg -format %[quality] info:
```

**Reply by FredWahl** [reply]
Sadly, it still gives me 93 which is the ”compression quality”. I tried these to versions
magick -ping "E:\test\00.jpg" -format %[quality] info:
magick -ping "E:\test\00.jpg" -format %[Q] info:
and they both return 93. If I understand documentation, the first should return the perceived quality of the jpg and the second the ”compression quality”, a metric I don’t understand. I found a discussion here https://superuser.com/questions/62730/how-to-find-the-jpg-quality where someone showed the source code that did what I expect. Imagemagick++ seems to have a method that gets this value but it just didn’t work for me. I got an exception as soon as it as given the file. Something about a missing handler. I found this tool https://www.mediachance.com/digicam/jpgq.htm which I’m pretty sure works but it would be nice to know how to do this in imagemagick. To reiterate, i successfully compressed a high quality jpg, several megabytes with cjpeg and 75 as quality so this number is what I want.

**Reply by FredWahl** [reply]
What a mess! I tried the exiftool on the original and it returned 93 which is what i expected. Then it dawned on me that I had run imagemagick on the original file, not the compressed. So, i wrotoe down the command in a text file to make sure it was right. When I ran the command, it returned 0. No quality at all. The tool i found at https://www.mediachance.com/digicam/jpgq.htm returned <Unknown> 
So I feel lost right now.

**Comment by fmw42** [comment]
Post your file so one of us can test with it

**Comment by FredWahl** [comment]
While doing research, i found this https://github.com/mozilla/mozjpeg/issues/307 and one of the comments comes from one of the developers of mozjpeg and he said ” ImageMagick doesn't calculate quality correctly. Ignore what ImageMagick says. Its measure of quality is a rough hack designed to guess old libjpeg and nothing else.” I now think that i ask for the right thing but the code doesn’t work. However, i found this on github https://gist.github.com/atr000/559324 and I think this works. I need to compile and if I get something close to 75 then I know for sure. Imagemagick returns 0. If it works, i will post some files.

**Reply by fmw42** [reply]
Please return with what you find.  If it is true that IM is using old metric, then we should have it updated. Post a bug report if so.

**Comment by FredWahl** [comment]
I have done some test and the results are in the commands.txt file in the zip file. I have also included a VS project. The results are inconclusive for some reasons. When I use the _rgb option when I compress the original, I get mostly green images. So I don''t know what can be said about quality since something has gone wrong. When I didn't use this option, I get images that look OK but I also get an error that says 
ERROR. Wrong size for quantization table -- this contains 258 bytes (2 bytes short or 63 bytes long)
when I use the code by Neal Krawetz. I think Neal got it right and that the results would have been close to what you expect. This should be enough for you to replicate this issue. I'm not sure if we can say there's a bug since the output files have issues

Below is what I documented and I wrote down the commands first and then pasted them to make sure i got it right

Imagemagick is in the C:\Program Files\ImageMagick-7.1.1-Q16-HDRI folder which should indicate version
cjpeg is in the C:\\mozjpeg-v4.0.3-win-x64\\static\\Release\\cjpeg-static.exe folder which should indicate version
Windows version 10.0.19045.4412
Microsoft Visual Studio Professional 2022 (64-bit) - Current
Version 17.6.3
My VmWare partition is full and I can't upgrade VS until I have given it more space

he test image comes from https://getsamplefiles.com/sample-image-files/jpg#google_vignette and it was the sample-1.jpg
that I downloaded and then renamed to Original.jpg. This may not have been the best image to start with and if you
want to replicate, you should probably use a high quality image guaranteed to be unadultered.
The -rgb option used in cjpeg resulted in strange images where green is the dominant color.


"C:\mozjpeg-v4.0.3-win-x64\static\Release\cjpeg-static.exe" -quality 75 -rgb -optimize -sample 1x1 -progressive -outfile "E:\Test Images\rgbQ75.jpg" "E:\Test Images\Original.jpg"
"C:\mozjpeg-v4.0.3-win-x64\static\Release\cjpeg-static.exe" -quality 65 -rgb -optimize -sample 1x1 -progressive -outfile "E:\Test Images\rgbQ65.jpg" "E:\Test Images\Original.jpg"
"C:\mozjpeg-v4.0.3-win-x64\static\Release\cjpeg-static.exe" -quality 50 -rgb -optimize -sample 1x1 -progressive -outfile "E:\Test Images\rgbQ50.jpg" "E:\Test Images\Original.jpg"
"C:\mozjpeg-v4.0.3-win-x64\static\Release\cjpeg-static.exe" -quality 45 -rgb -optimize -sample 1x1 -progressive -outfile "E:\Test Images\rgbQ45.jpg" "E:\Test Images\Original.jpg"


"C:\mozjpeg-v4.0.3-win-x64\static\Release\cjpeg-static.exe" -quality 75 -optimize -sample 1x1 -progressive -outfile "E:\Test Images\Q75.jpg" "E:\Test Images\Original.jpg"
"C:\mozjpeg-v4.0.3-win-x64\static\Release\cjpeg-static.exe" -quality 65 -optimize -sample 1x1 -progressive -outfile "E:\Test Images\Q65.jpg" "E:\Test Images\Original.jpg"
"C:\mozjpeg-v4.0.3-win-x64\static\Release\cjpeg-static.exe" -quality 50 -optimize -sample 1x1 -progressive -outfile "E:\Test Images\Q50.jpg" "E:\Test Images\Original.jpg"
"C:\mozjpeg-v4.0.3-win-x64\static\Release\cjpeg-static.exe" -quality 45 -optimize -sample 1x1 -progressive -outfile "E:\Test Images\Q45.jpg" "E:\Test Images\Original.jpg"


Magick identify -format "%Q" "E:\Test Images\Original.jpg"  -> 73
Magick identify -format "%Q" "E:\Test Images\rgbQ75.jpg"    -> 62 At least it's lower than 73
Magick identify -format "%Q" "E:\Test Images\rgbQ65.jpg"    -> 92 Very unexpected
Magick identify -format "%Q" "E:\Test Images\rgbQ50.jpg"    -> 92 Very unexpected
Magick identify -format "%Q" "E:\Test Images\rgbQ45.jpg"    -> 92 Very unexpected

Magick identify -format "%Q" "E:\Test Images\Q75.jpg"    -> 92 Very unexpected
Magick identify -format "%Q" "E:\Test Images\Q65.jpg"    -> 92 Very unexpected
Magick identify -format "%Q" "E:\Test Images\Q50.jpg"    -> 92 Very unexpected
Magick identify -format "%Q" "E:\Test Images\Q45.jpg"    -> 92 Very unexpected

About the files. Using the -rgb option makes the output files mostly green
Original.jpg 1062185 bytes
rgbQ75.jpg 1445752 bytes. It's larger than the original and mostly green
rgbQ65.jpg 1190535 bytes. It's larger than the original and mostly green
rgbQ50.jpg 951912 bytes. It's smaller than the original and mostly green
rgbQ45.jpg 894018 byte. It's smaller than the original and mostly green

Q75.jpg 707394 bytes. Smaller than original and looks OK
Q65.jpg 575440 bytes. Smaller than original and looks OK
Q50.jpg 448024 bytes. Smaller than original and looks OK
Q45.jpg 420122 bytes. Smaller than original and looks OK

Now I run the code I found on https://gist.github.com/atr000/559324 on the mostly green files
E:\jpegquality.exe "E:\Test Images\Original.jpg" -> 79.84 This value does not agree with what I get from magick.
E:\jpegquality.exe "E:\Test Images\rgbQ75.jpg" -> 54.6 This is lower than 79.84 which is expected
E:\jpegquality.exe "E:\Test Images\rgbQ65.jpg" -> 52.4 This is lower than 54.6 which is expected
E:\jpegquality.exe "E:\Test Images\rgbQ50.jpg" -> 38.55 This is lower than 52.4 which is expected
E:\jpegquality.exe "E:\Test Images\rgbQ45.jpg" -> 38.66 Something fishy here, how can the quality be higher?

E:\jpegquality.exe "E:\Test Images\Q75.jpg" -> 54.6
E:\jpegquality.exe "E:\Test Images\Q65.jpg" -> 88.19 and says "ERROR. Wrong size for quantization table -- this contains 258 bytes (2 bytes short or 63 bytes long)
E:\jpegquality.exe "E:\Test Images\Q50.jpg" -> 83.17 and says "ERROR. Wrong size for quantization table -- this contains 258 bytes (2 bytes short or 63 bytes long)
E:\jpegquality.exe "E:\Test Images\Q45.jpg" -> 81.32 and says "ERROR. Wrong size for quantization table -- this contains 258 bytes (2 bytes short or 63 bytes long)" but it still calculates something



At least the quality returned is lower the more compressed they are. However, they are larger than what I get from the original file
I assume they would have decreased in size the more they had been compressed and also had a lower value than original had there not been a problem
with the quantization table(s) 




Imagemagick is in the C:\Program Files\ImageMagick-7.1.1-Q16-HDRI folder which should indicate version
cjpeg is in the C:\\mozjpeg-v4.0.3-win-x64\\static\\Release\\cjpeg-static.exe folder which should indicate version
Windows version 10.0.19045.4412
Microsoft Visual Studio Professional 2022 (64-bit) - Current
Version 17.6.3
My VmWare partition is full and I can't upgrade VS until I have given it more space

he test image comes from https://getsamplefiles.com/sample-image-files/jpg#google_vignette and it was the sample-1.jpg
that I downloaded and then renamed to Original.jpg. This may not have been the best image to start with and if you
want to replicate, you should probably use a high quality image guaranteed to be unadultered.
The -rgb option used in cjpeg resulted in strange images where green is the dominant color.


"C:\mozjpeg-v4.0.3-win-x64\static\Release\cjpeg-static.exe" -quality 75 -rgb -optimize -sample 1x1 -progressive -outfile "E:\Test Images\rgbQ75.jpg" "E:\Test Images\Original.jpg"
"C:\mozjpeg-v4.0.3-win-x64\static\Release\cjpeg-static.exe" -quality 65 -rgb -optimize -sample 1x1 -progressive -outfile "E:\Test Images\rgbQ65.jpg" "E:\Test Images\Original.jpg"
"C:\mozjpeg-v4.0.3-win-x64\static\Release\cjpeg-static.exe" -quality 50 -rgb -optimize -sample 1x1 -progressive -outfile "E:\Test Images\rgbQ50.jpg" "E:\Test Images\Original.jpg"
"C:\mozjpeg-v4.0.3-win-x64\static\Release\cjpeg-static.exe" -quality 45 -rgb -optimize -sample 1x1 -progressive -outfile "E:\Test Images\rgbQ45.jpg" "E:\Test Images\Original.jpg"


"C:\mozjpeg-v4.0.3-win-x64\static\Release\cjpeg-static.exe" -quality 75 -optimize -sample 1x1 -progressive -outfile "E:\Test Images\Q75.jpg" "E:\Test Images\Original.jpg"
"C:\mozjpeg-v4.0.3-win-x64\static\Release\cjpeg-static.exe" -quality 65 -optimize -sample 1x1 -progressive -outfile "E:\Test Images\Q65.jpg" "E:\Test Images\Original.jpg"
"C:\mozjpeg-v4.0.3-win-x64\static\Release\cjpeg-static.exe" -quality 50 -optimize -sample 1x1 -progressive -outfile "E:\Test Images\Q50.jpg" "E:\Test Images\Original.jpg"
"C:\mozjpeg-v4.0.3-win-x64\static\Release\cjpeg-static.exe" -quality 45 -optimize -sample 1x1 -progressive -outfile "E:\Test Images\Q45.jpg" "E:\Test Images\Original.jpg"


Magick identify -format "%Q" "E:\Test Images\Original.jpg"  -> 73
Magick identify -format "%Q" "E:\Test Images\rgbQ75.jpg"    -> 62 At least it's lower than 73
Magick identify -format "%Q" "E:\Test Images\rgbQ65.jpg"    -> 92 Very unexpected
Magick identify -format "%Q" "E:\Test Images\rgbQ50.jpg"    -> 92 Very unexpected
Magick identify -format "%Q" "E:\Test Images\rgbQ45.jpg"    -> 92 Very unexpected

Magick identify -format "%Q" "E:\Test Images\Q75.jpg"    -> 92 Very unexpected
Magick identify -format "%Q" "E:\Test Images\Q65.jpg"    -> 92 Very unexpected
Magick identify -format "%Q" "E:\Test Images\Q50.jpg"    -> 92 Very unexpected
Magick identify -format "%Q" "E:\Test Images\Q45.jpg"    -> 92 Very unexpected

About the files. Using the -rgb option makes the output files mostly green
Original.jpg 1062185 bytes
rgbQ75.jpg 1445752 bytes. It's larger than the original and mostly green
rgbQ65.jpg 1190535 bytes. It's larger than the original and mostly green
rgbQ50.jpg 951912 bytes. It's smaller than the original and mostly green
rgbQ45.jpg 894018 byte. It's smaller than the original and mostly green

Q75.jpg 707394 bytes. Smaller than original and looks OK
Q65.jpg 575440 bytes. Smaller than original and looks OK
Q50.jpg 448024 bytes. Smaller than original and looks OK
Q45.jpg 420122 bytes. Smaller than original and looks OK

Now I run the code I found on https://gist.github.com/atr000/559324 on the mostly green files
E:\jpegquality.exe "E:\Test Images\Original.jpg" -> 79.84 This value does not agree with what I get from magick.
E:\jpegquality.exe "E:\Test Images\rgbQ75.jpg" -> 54.6 This is lower than 79.84 which is expected
E:\jpegquality.exe "E:\Test Images\rgbQ65.jpg" -> 52.4 This is lower than 54.6 which is expected
E:\jpegquality.exe "E:\Test Images\rgbQ50.jpg" -> 38.55 This is lower than 52.4 which is expected
E:\jpegquality.exe "E:\Test Images\rgbQ45.jpg" -> 38.66 Something fishy here, how can the quality be higher?

E:\jpegquality.exe "E:\Test Images\Q75.jpg" -> 54.6
E:\jpegquality.exe "E:\Test Images\Q65.jpg" -> 88.19 and says "ERROR. Wrong size for quantization table -- this contains 258 bytes (2 bytes short or 63 bytes long)
E:\jpegquality.exe "E:\Test Images\Q50.jpg" -> 83.17 and says "ERROR. Wrong size for quantization table -- this contains 258 bytes (2 bytes short or 63 bytes long)
E:\jpegquality.exe "E:\Test Images\Q45.jpg" -> 81.32 and says "ERROR. Wrong size for quantization table -- this contains 258 bytes (2 bytes short or 63 bytes long)

At least the quality returned is lower the more compressed they are. However, they are larger than what I get from the original file
I assume they would have decreased in size the more they had been compressed and also had a lower value than original had there not been a problem
with the quantization table(s) 
[Test Images.zip](https://github.com/user-attachments/files/15776547/Test.Images.zip)

**Reply by fmw42** [reply]
I am confused.  Are these "magick" commands?  Where are your "magick" commands to convert the image.

**Reply by FredWahl** [reply]
I assume you opened the zip file. I used cjpeg to compress files, magick.exe to get compression quality and then the code I found on github that had to be compiled. I did not use magick to convert images, all I wanted from magick was to get the quality value. The question is whether it does so ”correctly” and I think it doesn’t.

**Comment by FredWahl** [comment]
I did my tests on windows and one reply seemed to suggest that linux may yield a different result. I didn’t test with exiftool because the output files had issues. I didn’t run jpegtran on the original file, maybe I should have done so? At least, it has never produced funky output files. I should perhaps have included a file compressed by the commercial JPEGmini program. The problem with this program is that it offers no control whatsoever. It may or may not compress a file, you just don’t know.

**Reply by fmw42** [reply]
Thanks. I understand now.  So where did the -quality 75 arise, when IM said it was 93 or whatever value you found?

**Comment by FredWahl** [comment]
Alas, i didn’t use a high quality image to start with, i hoped the image I downloaded would be good enough. When I ran cjpeg, I used 75 as quality and assumed it was lower than what the original had. So the problem is that if I have a really good jpg and compress it using 75, i hope there would be a way to analyse it and get something close to 75. I have no idea how cjpeg understands this input value. My reason for getting this value is that I don’t want to my little utility to compress a jpg that already has been compressed. So if you get something low, i expect it has been processed somehow and then I do nothing. If it finds a large file and quality is high then it can. But it should never be recompressed more than once

**Comment by FredWahl** [comment]
Maybe my comment got lost? I thought that if I have a high quality image, ideally directly from my camera and magick thinks quality is 92, then if I compress it to make it smaller and use 75 as quality, then magick should return something in the neighbourhood of 75. Whatever it returns, it should return something lower if i compress the original and use 50 as quality.  And even lower if if i use cjpeg twice, 75 on the original and then 50 on the file I got when I used 75.

**Comment by snibgo** [comment]
> ... I don’t want to my little utility to compress a jpg that already has been compressed.

So, using a measure of bytes per pixel seems more logical than the loosely-defined "quality".

**Comment by FredWahl** [comment]
Yes, It seems I can’t completely trust the calculated quality. I need to examine this metric.i have many pictures, some are mine, some come from friends and some from the Internet. I suspect social media will do something to reduce their size. Some friends have photoshop and what not. If it doesn’t come from my camera, it has probably been adulterated somehow and there are many programs that can do this.
